### DISCUSSION TITLE
Power Limiter - strange behaviour?

### DISCUSSION THREAD

**Discussion by vagabond235** [hoylabs/OpenDTU-OnBattery#225] [Q&A] [answered]
Hallo - und herzlichen Dank für all die Mühe die in dieses Projekt fliesst! Ich stelle mal meine Frage auf Englisch, vielleicht hilfts der Nachwelt :)

I have openDTU-OnBattery running here in my setup of 4x420W panels (2s2p), 2 Victron 100/20 48V connected to a 48V no-name LiIon Battery and a Hoymiles HM-1500, and a Kaifa Smartmeter (Tinetz) to measure power - all interfacing via iobroker. So far i have not managed to get the Ve.Direct interface running - so currently i would just like to have openDTU setting power based on feedback from the Smartmeter and the battery voltage (as measured by the Hoymiles inverter).
Manually setting limits works just fine, just when i activate the power limiter, things go (mostly) wild - with the exception of high power demand (like the heatpump recharging hot water). It seems that the power limiter tries to set max power (with rather lots of power going back to grid) - and then losing connection to the inverter for too long, shutting down the inverter, before restarting. Under high load, this does not happen. Sometimes (but only shortly) the power limiter seems to work "normal".

"normal" load:
![image](https://user-images.githubusercontent.com/28185972/236698912-b4e13104-9d3a-43b3-af88-e6ed81beea7c.png)
 vs "high load"
![image](https://user-images.githubusercontent.com/28185972/236699277-c8e04544-49d1-4a7b-8b2d-47e1f06815ca.png)
vs expected behaviour:
![image](https://user-images.githubusercontent.com/28185972/236699419-4024cada-db71-44a1-a586-fb4b0b78df31.png)

Any ideas how to troubleshoot?
I have tried to play with the settings - increasing time of "Abfrageintervall" for DTU from 5s up to 15s, play with transmit power, power limiter settings... no significant changes
![image](https://user-images.githubusercontent.com/28185972/236699220-60109b35-c565-47b3-9c88-b0f406feb18b.png)

> PowerMeterClass: TotalPower:  0.00
21:46:44.587 > [PowerLimiterClass::loop] Power Meter/Inverter values too old. Using 0W (i.e. disable inverter)
21:46:44.587 > TX ActivePowerControl Channel: 3 --> 51 83 12 22 24 80 12 82 00 81 0B 00 00 64 00 00 BF 40 C7 
21:46:44.637 > Interrupt received
21:46:44.644 > RX Channel: 40 --> D1 83 12 22 24 83 12 22 24 81 00 00 0B 00 14 07 48 
21:46:46.635 > RX Period End
21:46:46.635 > Success
21:46:46.665 > TX PowerControl Channel: 23 --> 51 83 12 22 24 80 12 82 00 81 01 00 20 00 76 
21:46:48.203 > PowerMeterClass: TotalPower:  0.00
21:46:48.703 > RX Period End
21:46:48.703 > All missing
21:46:48.703 > Nothing received, resend whole request
21:46:48.703 > TX PowerControl Channel: 40 --> 51 83 12 22 24 80 12 82 00 81 01 00 20 00 76 
21:46:48.723 > Interrupt received
21:46:48.734 > RX Channel: 3 --> D1 83 12 22 24 83 12 22 24 81 00 00 01 00 B4 01 E4 
21:46:50.737 > RX Period End
21:46:50.737 > Success
21:46:50.737 > Fetch inverter: XXXXXXXXXXXXXXX
21:46:50.756 > TX RealTimeRunData Channel: 61 --> 15 83 12 22 24 80 12 82 00 80 0B 00 64 58 00 2A 00 00 00 00 00 00 00 00 85 1C 96 
21:46:50.784 > Interrupt received
21:46:50.793 > RX Channel: 3 --> 95 83 12 22 24 83 12 22 24 01 00 01 01 FE 00 02 00 04 00 09 00 14 00 00 13 D3 B1 
21:46:50.836 > Interrupt received
21:46:50.847 > RX Channel: 3 --> 95 83 12 22 24 83 12 22 24 02 00 00 13 DC 00 65 00 65 01 FE 00 01 00 03 00 07 A2 
21:46:50.886 > Interrupt received
21:46:50.896 > RX Channel: 3 --> 95 83 12 22 24 83 12 22 24 03 00 0F 00 00 14 01 00 00 13 B1 00 66 00 65 09 2C 08 
21:46:50.935 > Interrupt received
21:46:50.945 > RX Channel: 3 --> 95 83 12 22 24 83 12 22 24 84 13 85 00 00 00 D7 00 00 00 00 02 0F 00 34 DA DF 6C 
21:46:51.344 > RX Period End
21:46:51.344 > Success
21:46:51.396 > TX AlarmData Channel: 75 --> 15 83 12 22 24 80 12 82 00 80 11 00 64 58 00 2A 00 00 00 00 00 00 00 00 5F 07 4D 
21:46:51.403 > Interrupt received
21:46:51.412 > RX Channel: 40 --> 95 83 12 22 24 83 12 22 24 01 00 01 B0 01 00 01 62 E6 62 E6 00 00 00 00 70 7C 29 
21:46:51.420 > Interrupt received
21:46:51.428 > RX Channel: 75 --> 95 83 12 22 24 83 12 22 24 03 69 B3 00 00 00 00 70 7C 00 1D 69 E8 6A 04 00 00 B2 
21:46:51.652 > Interrupt received
21:46:51.660 > RX Channel: 61 --> 95 83 12 22 24 83 12 22 24 07 00 00 70 7C 00 27 6B 4A 6B 52 00 00 00 00 70 7C AD 
21:46:51.862 > Interrupt received
21:46:51.874 > RX Channel: 75 --> 95 83 12 22 24 83 12 22 24 0B 00 31 6C C4 6C CA 00 00 00 00 70 7C 00 33 6D 01 F2 
21:46:51.878 > Interrupt received
21:46:51.883 > RX Channel: 3 --> 95 83 12 22 24 83 12 22 24 8C 6D 11 00 00 00 00 7D F2 EA 
21:46:52.164 > RX Period End
21:46:52.164 > Middle missing
21:46:52.164 > Request retransmit: 2
21:46:52.164 > TX RequestFrame Channel: 3 --> 15 83 12 22 24 80 12 82 00 82 10 
21:46:52.215 > Interrupt received
21:46:52.221 > RX Channel: 61 --> 95 83 12 22 24 83 12 22 24 02 00 19 69 30 69 42 00 00 00 00 70 7C 00 1B 69 76 F4 
21:46:52.232 > RX Period End
21:46:52.232 > Middle missing
21:46:52.232 > Request retransmit: 4
21:46:52.232 > TX RequestFrame Channel: 23 --> 15 83 12 22 24 80 12 82 00 84 16 
21:46:52.248 > Interrupt received
21:46:52.258 > RX Channel: 40 --> 95 83 12 22 24 83 12 22 24 04 00 00 70 7C 00 1F 6A 33 6A 44 00 00 00 00 70 7C F9 
21:46:52.471 > RX Period End
21:46:52.471 > Middle missing
21:46:52.471 > Request retransmit: 5
21:46:52.471 > TX RequestFrame Channel: 40 --> 15 83 12 22 24 80 12 82 00 85 17 
21:46:52.678 > RX Period End
21:46:52.678 > Middle missing
21:46:52.678 > Request retransmit: 5
21:46:52.678 > TX RequestFrame Channel: 61 --> 15 83 12 22 24 80 12 82 00 85 17 
21:46:52.762 > RX Period End
21:46:52.762 > Middle missing
21:46:52.762 > Request retransmit: 5
21:46:52.762 > TX RequestFrame Channel: 75 --> 15 83 12 22 24 80 12 82 00 85 17 
21:46:53.086 > RX Period End
21:46:53.086 > Middle missing
21:46:53.086 > Retransmit timeout
21:46:54.622 > [PowerLimiterClass::loop] Power Meter/Inverter values too old. Using 0W (i.e. disable inverter)
21:46:54.622 > TX ActivePowerControl Channel: 3 --> 51 83 12 22 24 80 12 82 00 81 0B 00 00 64 00 00 BF 40 C7 
21:46:54.666 > Interrupt received
21:46:54.676 > RX Channel: 61 --> D1 83 12 22 24 83 12 22 24 81 00 00 0B 00 14 07 48 
21:46:56.670 > RX Period End
21:46:56.670 > Success
21:46:56.720 > TX PowerControl Channel: 23 --> 51 83 12 22 24 80 12 82 00 81 01 00 20 00 76 
21:46:56.727 > Interrupt received
21:46:56.732 > RX Channel: 61 --> D1 83 12 22 24 83 12 22 24 81 00 00 01 00 B4 01 E4 
21:46:58.308 > PowerMeterClass: TotalPower:  0.00
21:46:58.719 > RX Period End
21:46:58.720 > Success
21:47:00.972 > Received MQTT message on topic: kaifa/sensor/meter01_active_power_plus/state
21:47:00.972 > PowerMeterClass: TotalPower: 362.00
21:47:01.004 > Received MQTT message on topic: kaifa/sensor/meter01_active_power_plus/state
21:47:01.004 > PowerMeterClass: TotalPower: 362.00
21:47:04.658 > [PowerLimiterClass::loop] dcVoltage: 51.00 Voltage Start Threshold: 50.00 Voltage Stop Threshold: 49.00 inverter->isProducing(): 0
21:47:04.658 > [PowerLimiterClass::loop] newPowerLimit: 352
21:47:04.658 > [PowerLimiterClass::loop] Starting up inverter...
21:47:04.658 > [PowerLimiterClass::loop] Limit Non-Persistent: 352 W
21:47:04.658 > TX PowerControl Channel: 40 --> 51 83 12 22 24 80 12 82 00 81 00 00 B0 01 E6 
21:47:06.603 > RX Period End
21:47:06.603 > All missing
21:47:06.603 > Nothing received, resend whole request

**Comment by peff74** [comment]
First of all, I find the distance between the upper and lower power limit too narrow because when the inverter is switched on, the battery voltage drops. This can be as much as one volt.
Then I noticed in the log: 
[PowerLimiterClass::loop] Power Meter/Inverter values too old. Using 0W (i.e. disable inverter)
I would interpret this as that your smart meter did not read any data.

**Comment by vagabond235** [comment]
Thanks for the hint on MQTT - i think i understand the system a bit better now. It seems it works too well :)

When power consumption reaches 0 on the smartmeter - MQTT seems not to broadcast any updates any more - which could cause the timeouts. If i manually change power between net-from-grid to net-to-grid I see the following on the console:

With power from grid:

12:48:34.262 > RX Channel: 40 --> 95 83 12 22 24 83 12 22 24 81 00 01 00 AA 00 00 03 E8 FF FF FF FF 01 68 00 A2 9F 
12:48:34.371 > RX Period End
12:48:34.371 > Success
12:48:35.181 > Received MQTT message on topic: kaifa/sensor/meter01_active_power_plus/state
12:48:35.181 > PowerMeterClass: TotalPower: 154.00
12:48:35.181 > Received MQTT message on topic: kaifa/sensor/meter01_active_power_plus/state
12:48:35.181 > PowerMeterClass: TotalPower: 154.00
12:48:40.133 > Received MQTT message on topic: kaifa/sensor/meter01_active_power_plus/state
12:48:40.133 > PowerMeterClass: TotalPower: 153.00
12:48:40.190 > Received MQTT message on topic: kaifa/sensor/meter01_active_power_plus/state
12:48:40.250 > PowerMeterClass: TotalPower: 153.00
12:48:43.235 > PowerMeterClass: TotalPower: 153.00
12:48:45.025 > Received MQTT message on topic: kaifa/sensor/meter01_active_power_plus/state
12:48:45.140 > PowerMeterClass: TotalPower: 152.00
12:48:45.191 > Received MQTT message on topic: kaifa/sensor/meter01_active_power_plus/state
12:48:45.191 > PowerMeterClass: TotalPower: 152.00
12:48:47.764 > Fetch inverter: XXXXXXXXXXXXX
12:48:47.821 > TX RealTimeRunData Channel: 23 --> 15 83 12 22 24 80 12 82 00 80 0B 00 64 58 D3 8F 00 00 00 00 00 00 00 00 74 2F 22 
12:48:47.956 > Interrupt received
12:48:48.002 > RX Channel: 3 --> 95 83 12 22 24 83 12 22 24 01 00 01 02 11 00 7E 00 7F 02 9A 02 9F 00 00 15 2F B8 

No power from grid (but going to grid):
12:55:33.807 > RX Period End
12:55:33.807 > Success
12:55:34.292 > PowerMeterClass: TotalPower:  0.00
12:55:44.315 > PowerMeterClass: TotalPower:  0.00
12:55:47.878 > Fetch inverter: XXXXXXXXXXXXX
12:55:47.944 > TX RealTimeRunData Channel: 3 --> 15 83 12 22 24 80 12 82 00 80 0B 00 64 58 D5 33 00 00 00 00 00 00 00 00 AC AF C0

So - no MQTT messages received, MQTT update arrive once every minute --> might be a too long interval for the power limiter?

**Reply by vagabond235** [reply]
Something i still do not understand is why the power limiter goes to its max. value - even though its already feeding back power to the grid - as seen in my first image (green: production, blue: power going to grid, yellow: power from grid).

On the voltage: I assumed this is setting the voltage when the inverter will be stopped (lower value) - and restarted when charging (higher value) - am I wrong here?

**Comment by vagabond235** [comment]
I think i am getting closer. Changing "Angestrebter Netbezug" to 30W and "Hysterese für den Netzbezug" to 10 made the system more stable. However, a quick change in power need pushed the ESP back into its old behaviour:
![image](https://user-images.githubusercontent.com/28185972/236835959-1e2a5e29-9ab8-4c9d-91a9-da9843ee3208.png)

a (very bad) blockly in iobroker is able to get things back to normal:
![image](https://user-images.githubusercontent.com/28185972/236836282-dee8b1c3-eedf-44b1-95a7-d79e53bc14db.png)

it basically just manually (re-)sets the 0 from the power meter so it broadcasts a value to the ESP. 
![image](https://user-images.githubusercontent.com/28185972/236837049-9e695390-8f66-4a5b-9bbb-6bbe1a5cd8c9.png)

Would it be possible to add some sort of error handling for this case to the code of the power limiter? Like using the power-to-grid value (if available) to calculate the required power of the inverter.

**Comment by helgeerbe** [comment, marked as answer]
power limiter expects a positive value for power from grid and a negativ value for power to grid.
So why don't you send on one topic `"power from grid" - "power to grid"`, or on 2 different topics "power from grid" and on the second topic "power to grid * -1"

**Reply by vagabond235** [reply]
Thats what happens if 44 year old biologists turn script-kiddy :) - not seeing the obvious. 
Thanks a lot - could get it sorted this way. Had to tell my ESPhome sensor to do the math (multiply outgoing with -1) - as doing the math in iobroker crashed openDTU (whatever i did wrong there...)
